### PR TITLE
Tuakiri VHO User Enumeration Fix

### PR DESCRIPTION
**Pull request by Razorfang**
With this change, you can't visit the password reset page unless you've received an email link first. This means that attackers can't figure out who does or doesn't have an account by brute forcing usernames.

This change will also tell any user without a phone number that they should contact an administrator and get a phone number.

### CODE REVIEW

File: virtualhome/grails-app/views/lostPassword/emailed.gsp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  <body>
+
+    <div class="hero-unit">
+      <h2>An email has been sent to this user.</h2>
+
+      <g:render template="/templates/flash" plugin="aafApplicationBase"/>
+
+      <p>Please follow the instructions.</p>
+
+      <h3>Didn't receive an email?</p>
```
Comment (vladimir-mencl-eresearch) [inline virtualhome/grails-app/views/lostPassword/emailed.gsp]:
```suggestion
      <p>Please follow the instructions in the email.</p>

      <h3>Did not receive the email?</p>
```

**Comment by vladimir-mencl-eresearch** [comment]
Hi @Razorfang ,

Testing now - will provide more details here.

(1) Re-send email: thanks for making this a button.  Can you please change the way the username is passed from including it directly in the POST URL to having it as a (hidden) `input` instead?  This will avoid exposing it in browsing history ... and will generally look more robust.

(2) Just tried testing the flow for users without phone number.  I do get the email with a link, but when I follow the link, I only get the error message, not the dialog box to enter the SMS code (communicated out of band).


That's two things I managed to check now - more later.

Cheers,
Vlad
Screenshot 2024-11-04 at 16 20 16

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  I have addressed your two points.

File: virtualhome/grails-app/controllers/aaf/vhr/FinalizationController.groovy [hunk 2]
```diff
@@ -35,11 +35,24 @@ class FinalizationController {
...
       render "true"
+      return
+    }
+
+    if (managedSubjectInstance.id != id) {
+      render "false"
+      return
+    }
+
+    render "true"
```
Comment (vladimir-mencl-eresearch) [inline virtualhome/grails-app/controllers/aaf/vhr/FinalizationController.groovy]: I think the original logic flow (two branches, one returning `true`, one `false`) was easier to read.

The flow here leaves the final `render "true` as the edge-case which the reader of the code has to work out as "an account was found and it is the same account this invite is for").

How about changing it (almost back) to:
```suggestion
    if (managedSubjectInstance && managedSubjectInstance.id != id ) {
      render "false"
    }
    else
    {
      render "true"
    }
```

**Comment by vladimir-mencl-eresearch** [comment]
Thanks, the first two points are now addressed.

I'm now working my way through the code changes - and I started with the FinalizationController ... because it was easy to pick apart from the rest.

(3) I've just made a suggestion to the code style in the `loginAvailable` method.  But while looking there, I found an issue with the existing code and how it interacts with our changes to improve security.

The `index` method just above accepts the invite and puts it into the context.  It also checks whether the invite was already used, and stops with an error - but *only after putting the invite into the session context*.

So while a user who just visits an already used invite link gets an error message saying the invite was already used, the user's session gets the invite code blessing and is allowed to call the `loginAvailable` method (just tested it, indeed behaves this way).

This should have an easy fix - can you please move the `session.setAttribute(MANAGED_SUBJECT_ID,....)` line in the `index` method *after* the two checks it does, almost to the end of the method (just before it returns the values to be passed to the view).

I'll continue with the review - passing you what I find as I go.

**Comment by Razorfang** [comment]
> Thanks, the first two points are now addressed.
> 
> I'm now working my way through the code changes - and I started with the FinalizationController ... because it was easy to pick apart from the rest.

Done

**Comment by vladimir-mencl-eresearch** [comment]
Hi @Razorfang ,

I understand the scope of this work was initially a bit unclear and has grown over time - e.g. to also clean up the inconsistencies between email and SMS reset (and essentially undefined behaviour for `second_factor_required == false`).

Overall solid piece of work - though I still have some points to address.

I'll keep the numbering going up to avoid confusion with the previously raised points.

(4) I see that `emailed()` still sets CURRENT_USER in the session.  This really should be removed  - otherwise the user visiting the page (not necessarily the account owner) could just go to `/myaccount/reset` directly to (i) confirm the username exists (ii) proceed to the page bypassing email key check (but still needing SMS code).  The session key `CURRENT_USER` should only be set to confirm the user has provided a valid email code (in `obtainsubject()`).

(5) I see the `aaf.vhr.passwordreset.reset_code_length` config entry is used for both email and SMS code.  We want the email code reasonably long, but a SMS code one can type - I think this asks for two separate settings.

(6) There is some inconsistency in how resending impacts existing codes:
    1. Resending email code invalidates all previously emailed codes.   Either it should keep the code if already generated, or the wording should perhaps say that only latest code will work?
    2. Reloading the `reset()` page resends the code.  This might possibly result into some excessive sending.
    3. There is also the `resend()` function which: (i) is rate-limited to 5m, (ii) resends the *current* code, and (iii) redirects back to `reset()` which in turn generates and sends a new code.

I don't have a clear view on what the best behaviour  is, but right now it has strong potential for confusion, and possibly for flooding someone's inbox.


(7) In `virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp`, the reference to admins:
```
In no adminstrators are available for this account, please email <a href="mailto:tuakiri@reannz.co.nz">Tuakiri support at tuakiri@reannz.co.nz</a>
```
* has a typo in word administrators
* can look confusing, as if the recipient does not know that there might be admins configured for the account's organisation/group
I think the wording should be conditional on whether `${organizationAdminRole.subjects?.size() > 0 || ${groupAdminRole.subjects?.size() > 0` and say either:
`If you are unable to contact the above administrators, please email...`
or
`As your organisation and group does not have any administrators registered, please email...`



Now getting to boring code cleanup points:
(8) Have you tried running tests - do they still pass?  I liked the way `second_factor_required` was removed from across the application code, but it is still showing up in tests...

(9) `LostPasswordController` `allowedMethods`: restrict `emailed` to `POST` only - and `obtainsubject` does not need the (default) restriction to `GET`

(10) Possibly remove `sendResetCodes` method and call `sendsms` directly for clarity.  Calling it "codes" is now confusing... and it is now just a wrapper for sendsms with no added functionality.  (There is a check of return param, but not really effective, as SMSDeliveryService always returns true ... logic error in applicationbase.)

(11) `LostPasswordService`: if `sendResetEmail` is the public entry point, maybe the two methods called from there should be private?



The last two points (10, 11) are just optional suggestions.

(4) is quite high priority but easy to fix.

(6) may need some thinking - I'll be in the office tomorrow and we can discuss it than.


But great to see this taking shape well - should not be long to get it done.

Cheers,
Vlad

**Review by vladimir-mencl-eresearch — CHANGES_REQUESTED** [review]

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  Your proposal has been added

**Comment by vladimir-mencl-eresearch** [comment]
Second thoughts on (10) - perhaps instead fix the code in SMSDeliveryService to report SMS delivery failure and keep the code acting on the failure .... but this is at the level of polishing minor nits.

**Comment by Razorfang** [comment]
Point (4) addressed.

File: virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  <ul>
+    <g:each in="${organizationAdminRole.subjects?.sort{it.cn}}" var="subject">
+      <li><g:fieldValue bean="${subject}" field="cn"/> - <a href="mailto:${subject.email}"><g:fieldValue bean="${subject}" field="email"/></a></li>
+    </g:each>
+  </ul>
+</g:if>
+
+Once you have obtained the SMS code from your administrator, <a href="${emailURL}">click the link</a> to proceed with your password reset.<br><br>
+
+<g:if test="${groupAdminRole.subjects?.size() > 0}">
```
Comment (vladimir-mencl-eresearch) [inline virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp]: I think the condition may have to be an "or" over organizationAdminRole and groupAdminRole

Comment (Razorfang) [reply virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp]: Check the latest commit

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  If we end up changing `def code = aaf.vhr.crypto.CryptoUtil.randomAlphanumeric(128)` to make 128 a setting in our config file, I will have to open a merge request in the ansible repo. Just giving you a heads up.

**Comment by Razorfang** [comment]
> @vladimir-mencl-eresearch If we end up changing `def code = aaf.vhr.crypto.CryptoUtil.randomAlphanumeric(128)` to make 128 a setting in our config file, I will have to open a merge request in the ansible repo. Just giving you a heads up.

https://gogs.reannz.co.nz/REANNZ/ansible-playbooks/pulls/213

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  I just need to deal with point (6) now. I've noticed another issue.

Because resend() redirects to reset(), and we send the reset code every time reset() is visited, we actually send a code twice; we send the current code, then we generate a new code and send that one.

Here's how I will approach the issue:

Instead of sending the reset code when reset() is visited, we'll do it when obtainsubject() (the URL in the email) is clicked. This way it's only generated once since the page then redirects to reset(). This also prevents a regeneration and resend every time the page is reloaded.

File: virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp (hunk 3 above)
Comment (vladimir-mencl-eresearch) [reply virtualhome/src/groovy/aaf/vhr/email_password_reset_without_phone.gsp]: Great, thanks!

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  Everything besides points (10) and (11) should be covered now.

**Comment by vladimir-mencl-eresearch** [comment]
> @vladimir-mencl-eresearch If we end up changing `def code = aaf.vhr.crypto.CryptoUtil.randomAlphanumeric(128)` to make 128 a setting in our config file, I will have to open a merge request in the ansible repo. Just giving you a heads up.

Sure, no problem, I see you've [already started the work there](https://gogs.reannz.co.nz/REANNZ/ansible-playbooks/commit/43563183264c4e224ee26d2c905ae52272d09ab3).

**Comment by vladimir-mencl-eresearch** [comment]
Thanks @Razorfang for the patience with all my comments.

I like the creative fix for (6) - moving the sending to `obtainsubject()`.  That significantly improves the behaviour.

Clicking the email link now:
* Generates a new SMS code and sends the code once
* Resend button sends the code once more and redirects back to reset.
* Subsequent resends are rate-limited at 1 per 5m, but the first resend is free - that sounds OK.

Interesting edge-case behaviour is:
* Clicking the email link again invalides existing SMS code and sends a new one.  That sounds reasonable and secure.
* Resending the email also invalidates the email code.   Wondering whether the text on or around the Resend button should warn about this.  Your thoughts?

**Comment by vladimir-mencl-eresearch** [comment]
Btw, it might be good testing how the application handles failure of SMS message delivery.

The SMSDeliveryService in applicationbase has a logic error: at https://github.com/REANNZ/applicationbase/blob/master/grails-app/services/aaf/base/SMSDeliveryService.groovy#L14 `outcome` should be initialised to `false` to get changed to `true` only in case of success.

I have just pushed branch `fix-sms-delivery` to applicationbase and created a PR there: https://github.com/REANNZ/applicationbase/pull/1

Would you mind testing with a build made against this branch of applicationbase and intentionally breaking SMS delivery (changing both username and password in config file to nonsense) ?

Just to confirm what the code does if something goes wrong...

Cheers,
Vlad

**Comment by Razorfang** [comment]
>     * Resending the email also invalidates the email code.   Wondering whether the text on or around the Resend button should warn about this.  Your thoughts?

Having a warning seems fair. Although the current implementation is secure, I could imagine someone getting annoyed if they do things out of sequence. I'll make one last update.

**Comment by Razorfang** [comment]
Actually, I think it would make sense to change the text from 're-send email' to 'Send another email', so that people understand that we're not sending the same email again.

**Comment by vladimir-mencl-eresearch** [comment]
> Actually, I think it would make sense to change the text from 're-send email' to 'Send another email', so that people understand that we're not sending the same email again.

Maybe `Send a new code` ?

**Comment by Razorfang** [comment]
No I think we should say `email` instead of `code`. We're sending an email, which contains a code.

**Comment by vladimir-mencl-eresearch** [comment]
OK - make sense, thanks.

**Comment by vladimir-mencl-eresearch** [comment]
Btw, I've just tried on tuakiri-vho-dev what happens with the current code (and fixed SMSDeliveryService) if SMS delivery fails.

On `resend`, I got a blank page returned with just a `500`, but no graceful error page, just completely blank.

And in `app-stacktrace.log`, I saw a looong stacktrace starting with:
```
org.codehaus.groovy.grails.web.servlet.mvc.exceptions.CannotRedirectException: Cannot issue a redirect(..) here. A previous call to redirect(..) has already redirected the response.
```

I think it would be worth testing in the new code....

File: virtualhome/grails-app/views/lostPassword/emailed.gsp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+      <h3>Didn't receive the email?</p>
+
+      <g:form action="emailed">
+        <g:hiddenField name="login" value="${login}"/>
+        <button type="submit" class="btn btn-large">Send another email.</button>
+        <g:link controller="lostpassword"></g:link>
+      </g:form>
+
+      <h3><strong>NOTE: </strong>Only the most recently sent email will contain a vaild URL</h3>
```
Comment (vladimir-mencl-eresearch) [inline virtualhome/grails-app/views/lostPassword/emailed.gsp]: Sorry, a typo in one word ("vaild") ... and suggesting a slight rewording, plus using the style from getbootstrap used elsewhere for warning notices:
```suggestion
      <div class="alert alert-block alert-warning"><h4>Note</h4>Only the URL from the most recent email will be valid.</div>
```

Attaching the rendered result (played with browser tools in F
Screenshot 2024-11-07 at 08 57 04
irefox)

**Comment by Razorfang** [comment]
@vladimir-mencl-eresearch  Everything should be good to go now

**Comment by vladimir-mencl-eresearch** [comment]
Thanks - yes, all good to merge now.

**Review by vladimir-mencl-eresearch — APPROVED** [review]

**Pull request merged**